codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c

```diff
@@ -0,0 +1,322 @@
+typedef struct {
+  int *value;
+  size_t used;
+  size_t size;
+} qflgData;
+
+
+void free_parameters(struct RadarParm *prm, struct FitData *fit, FILE *fp, qflgData *q)
+{
+  if (prm != NULL) RadarParmFree(prm);
+  if (fit != NULL) FitFree(fit);
+  if (fp !=NULL) free(fp);
+  if (q->value !=NULL) {
+    free(q->value);
+    q->value=NULL;
+  }
+}
+
+
+
+int main (int argc,char *argv[]) {
+  
+  unsigned char vb=0;
+  unsigned char help=0;
+  unsigned char option=0;
+  unsigned char version=0;
+
+  OptionAdd(&opt,"-help",'x',&help);
+  OptionAdd(&opt,"-option",'x',&option);
+  OptionAdd(&opt,"-version",'x',&version);
+  OptionAdd(&opt,"vb",'x',&vb);
+
+  int arg;
+  arg=OptionProcess(1,argc,argv,&opt,rst_opterr);
+
+  if (arg==-1) {
+    exit(-1);
+  }
+
+  if (help==1) {
+    OptionPrintInfo(stdout,hlpstr);
+    exit(0);
+  }
+  if (option==1) {
+    OptionDump(stdout,&opt);
+    exit(0);
+  }
+
+  if (version==1) {
+    OptionVersion(stdout);
+    exit(0);
+  }
+  
+  FILE *fp=NULL;
+  if (arg==argc) fp=stdin;
+  else fp=fopen(argv[arg],"r");
+  if (fp==NULL) {
+    fprintf(stderr,"File not found.\n");
+    exit(-1);
+  }
+  
+  //set origin.command field
+  time_t ctime;
+  int c,n;
+  char command[128];
+  char tmstr[40];
+  command[0]=0;
+  n=0;
+  for (c=0;c<argc;c++) {
+    n+=strlen(argv[c])+1;
+    if (n>127) break;
+    if (c !=0) strcat(command," ");
+    strcat(command,argv[c]);
+  }
+  
+  
+  prm=RadarParmMake();
+  fit=FitMake();
+  if (FitFread(fp,prm,fit)==-1) {
+    fprintf(stderr,"Error reading file\n");
+    exit(-1);
+  }
+
+  // Initial memory allocation to store qflgs
+  qflgData *qflgs=malloc(sizeof(qflgData));
+  qflgs->value = malloc(tmax*maxrng*maxch*maxbm * sizeof(int));
+  qflgs->used = 0;
+  qflgs->size = tmax*maxrng*maxch*maxbm;
+  
+  
+  // Read the FITACF file to populate the qflg array
+  int index;
+  int maxindex=-1;
+  int rng,bm,ch;
+  int nrec[maxbm][maxch]; // counts number of records for each beam/channel
+  memset(nrec,0,sizeof(nrec));
+  
+  do {
+  
+    bm=prm->bmnum;
+    ch=prm->channel;
+    for (rng=0;rng<prm->nrang;rng++) {
+      index=get_index(bm,ch,rng,nrec[bm][ch],maxbm,maxch,maxrng);
+      if (maxindex < index) maxindex=index;
+      if (qflgs->size < maxindex) {
+        qflgs->size += tmax*maxrng*maxch*maxbm;
+        qflgs->value = realloc(qflgs->value, qflgs->size * sizeof(int));
+      }
+      qflgs->value[index] = fit->rng[rng].qflg;
+      qflgs->used = maxindex;
+      
+    }
+    nrec[bm][ch]++;
+  
+  } while (FitFread(fp,prm,fit) !=-1);
+  
+  
+  // Do the median filtering
+  //   Since qflg can only be 0 or 1, the median qflg of the 3x3 grid can be 
+  //     calculated using the test ( sum_of_qflgs >= 5 ).
+  //   Replicate padding is used to handle corner/edge cases
+  int sum;
+  int index_list[9];  
+  memset(index_list,-1,sizeof(index_list));
+  
+  
+  //** Read the file again, apply the median filter, and write a new file
+  // rewind the file pointer
+  rewind(fp);
```

Comment:
Does rewind return anything?

Reply:
I don't think so. This is how it's used elsewhere in RST, so I just copied it.